File: NeetCode/71/71.cpp
class Solution {

private:
    vector<vector<int>> res;

    void backtrack(vector<int>& perm, vector<int>& nums, vector<bool>& check){
        if(perm.size() == nums.size()){
            res.push_back(perm);
            return;
        }

        for(int i = 0; i < nums.size(); ++i){
            if (!check.at(i))
            {
                perm.push_back(nums.at(i));
                check.at(i) = true;
                backtrack(perm,nums,check);
                perm.pop_back();
                check.at(i) = false;
            }

        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> check(nums.size(), false);
        vector<int> perm;
        backtrack(perm, nums, check);
        return res;
    }

};

File: NeetCode/71/71.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.backtrack([], nums, [False] * len(nums))
        return self.res

    def backtrack(self, perm: List[int], nums: List[int], check: List[int]) -> None:
        if len(perm) == len(nums):
            print(perm)
            self.res.append(perm[:])
            return

        for i in range(len(nums)):
            if not check[i]:
                perm.append(nums[i])
                check[i] = True
                self.backtrack(perm, nums, check)
                perm.pop()
                check[i] = False